### ISSUE TITLE
helm fetch ignores build metadata in --version

### ISSUE THREAD

**Issue by borntyping** [helm/helm#4838]
When running `helm fetch`, incorrect charts are returned for version numbers containing build metadata. The build metadata segment appears to be entirely ignored, both when it matches a chart version and when it does not.

```bash
$ helm search darkroom -l | grep 5.0.0
example/example	5.0.0+e4201a3a        	1.0.0       	Chart description
example/example	5.0.0+bba96fca        	1.0.0       	Chart description
example/example	5.0.0+70350c25        	1.0.0       	Chart description

$ # Fetch a specific version - result does not match!
$ helm fetch example/example --version 5.0.0+70350c25
$ ls
example-5.0.0+bba96fca.tgz

$ # Fetch a non-existent version - should raise an error!
$ helm fetch example/example --version 5.0.0+XXXXXXX
$ ls
example-5.0.0+bba96fca.tgz
```

Output of `helm version`:

```
Client: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.11.0", GitCommit:"2e55dbe1fdb5fdb96b75ff144a339489417b146b", GitTreeState:"clean"}
```

**Comment by mattfarina** [comment]
This is clearly a bug. Thanks for finding an issue with `helm fetch`.

My quick guess as to what's happening is that the build metadata is being ignored early in the process.

According to the [Semantic Version spec](https://semver.org/):

> Build metadata SHOULD be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence.

Once that is ignored it tries to find a version (ignoring built metadata) that works.

PR's welcome.

**Comment by bobby-vandiver** [comment]
Bumping this issue since I recently ran into this and a related problem.

When running `helm dependency update`, if the version of a subchart includes build metadata and it is not the first chart version listed in the repository index, the correct version will not be retrieved and packaged. I think I've narrowed down the problem to this piece of code:
https://github.com/helm/helm/blob/master/pkg/resolver/resolver.go#L89

I believe you could workaround this limitation by doing surgery to the appropriate index file in Helm's cache to remove all versions and including the `--skip-refresh` option when running `helm dependency update`. Obviously you would need to inspect not only the target chart's requirements but all of the charts in the dependency graph and hack up the index file accordingly. Disclaimer: I have not tested this workaround.

Until this is resolved, I have switched to using the prerelease field of the chart version to encode the build metadata. This is not ideal, but is sufficient for my organization's needs.

**Comment by bacongobbler** [comment]
closed via #4850.

**Comment by noam-toga** [comment]
this still happens in helm 3.0.2 i've opened an issue here - 
https://github.com/helm/helm/issues/7563
